README.md
[![JetBrains Research](https://jb.gg/badges/research.svg)](https://confluence.jetbrains.com/display/ALL/JetBrains+on+GitHub)
Tests: Linux [![Tests Status](http://teamcity.jetbrains.com/app/rest/builds/buildType:(id:Epigenome_Tools_Npy)/statusIcon.svg)](http://teamcity.jetbrains.com/viewType.html?buildTypeId=Epigenome_Tools_Npy&guest=1)
Windows [![Build Status](http://teamcity.jetbrains.com/app/rest/builds/buildType:(id:Epigenome_Tools_Npy_NpyWindows)/statusIcon.svg)](http://teamcity.jetbrains.com/viewType.html?buildTypeId=Epigenome_Tools_Npy_NpyWindows&guest=1)

npy
===

`npy` allows to read and write files in [NPY] [npy] and [NPZ] [npy] formats
on the JVM.

[npy]: http://docs.scipy.org/doc/numpy-dev/neps/npy-format.html

Installation
------------

The latest version of `npy` is available on [jCenter] [jcenter]. If you're using
Gradle just add the following to your `build.gradle`:

```gradle
repositories {
    jcenter()
}

dependencies {
    compile 'org.jetbrains.bio:npy:0.3.3'
}

```

Examples
--------

### NPY

```kotlin
val values = intArrayOf(1, 2, 3, 4, 5, 6)
val path = Paths.get("sample.npy")
NpyFile.write(path, values, shape = intArrayOf(2, 3))

println(NpyFile.read(path))
// => NpyArray{data=[1, 2, 3, 4, 5, 6], shape=[2, 3]}
```

### NPZ

```kotlin
val values1 = intArrayOf(1, 2, 3, 4, 5, 6)
val values2 = booleanArrayOf(true, false)
val path = Paths.get("sample.npz")

NpzFile.write(path).use {
    it.write("xs", values1, shape = intArrayOf(2, 3))
    it.write("mask", values2)
}

NpzFile.read(path).use {
    println(it.introspect())
    // => [NpzEntry{name=xs, type=int, shape=[2, 3]},
    //     NpzEntry{name=mask, type=boolean, shape=[2]}]

    println("xs   = ${it["xs"]}")
    println("mask = ${it["mask"]}")
    // => xs   = NpyArray{data=[1, 2, 3, 4, 5, 6], shape=[2, 3]}
    //    mask = NpyArray{data=[true, false], shape=[2]}
}
```

Limitations
-----------

The implementation is rather minimal at the moment. Specifically it does
**not** support the following types:

* unsigned integral types (treated as signed),
* bit field,
* complex,
* object,
* Unicode
* void*
* intersections aka types for structured arrays.

[jcenter]: https://bintray.com/bintray/jcenter

Building from source
--------------------

The build process is as simple as

```bash
$ ./gradlew assemble
```

Testing
-------

No extra configuration is required for running the tests from Gradle

```bash
$ ./gradlew test
```

However, some tests require Python and NumPy to run and will be skipped
unless you have these.

Publishing
----------

You can publish a new release with a one-liner

```bash
./gradlew clean assemble test generatePomFileForMavenJavaPublication bintrayUpload
```

Make sure to set Bintray credentials (see API key section
[here](https://bintray.com/profile/edit)) in `$HOME/.gradle/gradle.properties`.

```
$ cat $HOME/.gradle/gradle.properties
bintrayUser=CHANGEME
bintrayKey=CHANGEME
```


CHANGES
npy Changelog
=============

Here you can see the full list of changes between each npy release.

Version 0.3.5
-------------

Bugfix release, released on January 14th, 2021

- Many NpzFile.Writer methods were made private in the previous release,
  breaking the API. This release reverts this change.


Version 0.3.4
-------------

Released on January 14th, 2021

- Fixed a bug with reading NPY files with `long` dimensions produced by
  Python 2.X. Thanks to @roblovelock! See PR #6 on GitHub.
- Fixed dependencies. The wildcard version specification made the library
  unusable in Maven, see issue #8 on GitHub.
- Bumped Kotlin version to 1.4.21, replaced deprecated methods.

Version 0.3.3
-------------

Bugfix release, released on September 28th, 2016

- Fixed tests on Windows.
- Fixed reading and writing string arrays.

Version 0.3.2
-------------

Released on September 23rd, 2016

- Fixed reading large (> 2G) NPY and NPZ files.
- Ensured the produced NPY file has no trailing gibberish.
- Changed `NpyFile.write` to automatically choose the appropriate
  format version.

Version 0.3.1
-------------

Released on September 21th, 2016

- Changed `NpyFile.write` and `NpzFile.Writer.write` to work in
  constant amount of memory.
- Allowed writing arrays using a non-native byte order.
- Dropped Guava dependency.

Version 0.3.0
-------------

Released on September 20th, 2016

- Added support for reading and writing arbitrary-dimensional arrays.

Version 0.2.1
-------------

Bugfix release, released on March 23rd, 2016

- Fixed metadata in NPY files created by `NpyFile.write`. See
  issue #4 on GitHub.

Version 0.2.0
-------------

Released on February 17th, 2016

- Unified reading and writing API of `NpzFile`.
- Added `NpzFile.Reader.introspect` for exploring the metadata of an
  NPZ file without reading whole arrays.
- Fixed `NpyFile` behaviour for arrays with strings of different lengths.

Version 0.1.0
-------------

Initial release, released on February 3rd, 2016

- Sketched reading and writing API. This is not yet finalized and can
  change in the future releases.
- Only 1D non-structured arrays of types 'b', 'i', 'u', 'f' and 'S'
  are supported.


src/test/kotlin/org/jetbrains/bio/npy/Examples.kt
package org.jetbrains.bio.npy

import java.nio.file.Path
import java.nio.file.Paths

internal object Examples {
    @JvmStatic operator fun get(name: String): Path {
        val url = Examples.javaClass.classLoader.getResource(name)
                  ?: throw IllegalStateException("resource not found")

        return Paths.get(url.toURI()).toFile().toPath()
    }
}

src/test/kotlin/org/jetbrains/bio/npy/NpyFileTest.kt
package org.jetbrains.bio.npy

import org.junit.Assert.assertArrayEquals
import org.junit.Assume
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import java.io.IOException
import java.lang.ProcessBuilder.Redirect
import java.nio.ByteOrder
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@RunWith(Parameterized::class)
class NpyFileTest(private val order: ByteOrder) {
    @Test fun writeReadBooleans() = withTempFile("test", ".npy") { path ->
        val data = booleanArrayOf(true, true, true, false)
        NpyFile.write(path, data)
        assertArrayEquals(data, NpyFile.read(path).asBooleanArray())
    }

    @Test fun writeReadBytes() = withTempFile("test", ".npy") { path ->
        val data = byteArrayOf(1, 2, 3, 4)
        NpyFile.write(path, data)
        assertArrayEquals(data, NpyFile.read(path).asByteArray())
    }

    @Test fun writeReadShorts() = withTempFile("test", ".npy") { path ->
        val data = shortArrayOf(1, 2, 3, 4)
        NpyFile.write(path, data, order = order)
        assertArrayEquals(data, NpyFile.read(path).asShortArray())
    }

    @Test fun writeReadInts() = withTempFile("test", ".npy") { path ->
        val data = intArrayOf(1, 2, 3, 4)
        NpyFile.write(path, data, order = order)
        assertArrayEquals(data, NpyFile.read(path).asIntArray())
    }

    @Test fun writeReadLongs() = withTempFile("test", ".npy") { path ->
        val data = longArrayOf(1, 2, 3, 4)
        NpyFile.write(path, data, order = order)
        assertArrayEquals(data, NpyFile.read(path).asLongArray())
    }

    @Test fun writeReadFloats() = withTempFile("test", ".npy") { path ->
        val data = floatArrayOf(1f, 2f, 3f, 4f)
        NpyFile.write(path, data, order = order)
        assertArrayEquals(data, NpyFile.read(path).asFloatArray(), Math.ulp(1f))
    }

    @Test fun writeReadDoubles() = withTempFile("test", ".npy") { path ->
        val data = doubleArrayOf(1.0, 2.0, 3.0, 4.0)
        NpyFile.write(path, data, order = order)
        assertArrayEquals(data, NpyFile.read(path).asDoubleArray(), Math.ulp(1.0))
    }

    @Suppress("unchecked_cast")
    @Test fun writeReadStrings() = withTempFile("test", ".npy") { path ->
        val data = arrayOf("foo", "bar", "bazooka")
        NpyFile.write(path, data)
        assertArrayEquals(data, NpyFile.read(path).asStringArray())
    }

    companion object {
        @JvmStatic
        @Parameters(name = "{0}")
        fun `data`(): Collection<Any> = listOf(ByteOrder.BIG_ENDIAN,
                                               ByteOrder.LITTLE_ENDIAN)
    }
}

@RunWith(Parameterized::class)
class NpyFileNDTest(private val shape: IntArray) {
    @Test fun writeReadDouble() = withTempFile("test", ".npy") { path ->
        val data = doubleArrayOf(1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0)
        NpyFile.write(path, data, shape = shape)
        val serialized = NpyFile.read(path)
        assertArrayEquals(shape, serialized.shape)
        assertArrayEquals(data, serialized.asDoubleArray(), Math.ulp(1.0))
    }

    companion object {
        @JvmStatic @Parameters fun `data`() = listOf(
                intArrayOf(1, 8),
                intArrayOf(8, 1),
                intArrayOf(2, 4),
                intArrayOf(4, 2),
                intArrayOf(2, 2, 2),
                intArrayOf(2, 2, 2, 1))
    }
}

class NpyFileHeaderTest {
    @Test fun isPadded() {
        val header = NpyFile.Header(type = 'i', bytes = 4, shape = intArrayOf(42))
        assertEquals(1, header.major)
        assertTrue(header.allocate().capacity() % 16 == 0)
    }

    @Test fun writeRead10() = testWriteRead(65536)

    @Test fun writeRead20() = testWriteRead(16)  // force 2.0

    private fun testWriteRead(boundary: Int = NpyFile.Header.NPY_10_20_SIZE_BOUNDARY) {
        withTempFile("test", ".npz") { path ->
            val backup = NpyFile.Header.NPY_10_20_SIZE_BOUNDARY
            NpyFile.Header.NPY_10_20_SIZE_BOUNDARY = boundary
            val header = NpyFile.Header(type = 'i', bytes = 4, shape = intArrayOf(0))
            NpyFile.Header.NPY_10_20_SIZE_BOUNDARY = backup

            FileChannel.open(path, StandardOpenOption.WRITE).use {
                val output = header.allocate()
                output.rewind()
                it.write(output)
            }

            FileChannel.open(path).use {
                val input = it.map(FileChannel.MapMode.READ_ONLY, 0, Files.size(path))
                assertEquals(header, NpyFile.Header.read(input))
            }
        }
    }


}

class NpyFileNumPyTest {
    private val hasNumPy: Boolean get() = try {
        val (rc, _) = command("python", "-c", "import numpy")
        rc == 0
    } catch(e: IOException) {
        if (e.message?.startsWith("Cannot run program \"python\"") != true) {
            throw e
        }
        // no python installed
        false
    }

    @Test fun writeRead() {
        Assume.assumeTrue(hasNumPy)

        withTempFile("test", ".npy") { path ->
            val data = intArrayOf(1, 2, 3, 4)
            NpyFile.write(path, data)
            val (rc, output) = command(
                    "python", "-c", "import numpy as np; print(np.load('$path'))")
            assertEquals(0, rc)
            assertEquals("[1 2 3 4]", output.trim())
        }
    }

    private fun command(vararg args: String): Pair<Int, String> {
        val p = ProcessBuilder()
                .command(*args)
                .redirectOutput(Redirect.PIPE)
                .start()

        val rc = p.waitFor()
        return rc to p.inputStream.bufferedReader().readText()
    }
}

class NpyFileStressTest {
    @Test(expected = IllegalStateException::class)
    fun readRandomGibberish() = withTempFile("test", ".npy") { path ->
        val r = Random()
        Files.write(path, ByteArray(65536) { r.nextInt().toByte() })
        NpyFile.read(path)
    }

    @Test fun writeReadRandom() {
        val r = Random()
        val maxMemory = Runtime.getRuntime().maxMemory() / java.lang.Double.BYTES
        val maxSize = Math.toIntExact(maxMemory / 100)  // 1%
        for (i in 0 until 10) {
            val size = r.nextInt(maxSize).toLong()
            val data = r.doubles(size).toArray()

            withTempFile("test", ".npy") { path ->
                NpyFile.write(path, data)

                assertArrayEquals(data, NpyFile.read(path).asDoubleArray(),
                                  Math.ulp(1.0))
            }
        }
    }

    @Test fun writeReadUnaligned() = withTempFile("test", ".npy") { path ->
        val data = Random().doubles(65536).toArray()
        NpyFile.write(path, data)
        assertArrayEquals(data, NpyFile.read(path, step=123).asDoubleArray(),
                          Math.ulp(1.0))
    }
}

class NpyFileNonRegressionTest {
    @Test fun readLongShapePython2() {
        // See https://github.com/JetBrains-Research/npy/pull/6.
        assertArrayEquals(
                doubleArrayOf(1.0, 1.0),
                NpyFile.read(Examples["win64python2.npy"]).asDoubleArray(),
                Math.ulp(1.0))
    }
}


src/test/kotlin/org/jetbrains/bio/npy/NpzFileTest.kt
package org.jetbrains.bio.npy

import org.junit.Assert.assertArrayEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import java.nio.ByteOrder
import java.util.*
import kotlin.test.assertEquals

/**
 * The test file was generated by the following Python script:
 *
 *     import numpy as np
 *
 *     np.savez("example.npz",
 *              x_i1=np.array([1, 2, 3, 4], dtype="|i1"),
 *              x_i2=np.array([1, 2, 3, 4, 4096], dtype=">i2"),
 *              x_u4=np.array([1, 2, 3, 4], dtype=">u4"),
 *              x_i8=np.array([1, 2, 3, 4, 65536], dtype="<i8"),
 *              x_f4=np.array([1, 2, 3, 4], dtype="<f4"),
 *              x_f8=np.array([1, 2, 3, 4], dtype="<f8"),
 *              x_b=np.array([True, True, True, False], dtype="|b1"),
 *              x_S3=np.array(["aha", "hah"], dtype="|S3"))
 *
 * NumPy version 1.10.2.
 */
class NpzFileTest {
    @Test fun introspectTypes() {
        NpzFile.read(Examples["example.npz"]).use { npzf ->
            assertEquals(mapOf("x_i1" to Byte::class.java,
                               "x_i2" to Short::class.java,
                               "x_u4" to Int::class.java,
                               "x_i8" to Long::class.java,
                               "x_f4" to Float::class.java,
                               "x_f8" to Double::class.java,
                               "x_b" to Boolean::class.java,
                               "x_S3" to String::class.java),
                         npzf.introspect().associateBy({ it.name }, { it.type }))
        }
    }

    @Test fun introspectNames() {
        NpzFile.read(Examples["example.npz"]).use { npzf ->
            assertEquals(setOf("x_i1", "x_i2", "x_u4", "x_i8",
                               "x_f4", "x_f8", "x_b", "x_S3"),
                         npzf.introspect().map { it.name }.toSet())
        }
    }

    @Suppress("unchecked_cast")
    @Test fun get() {
        NpzFile.read(Examples["example.npz"]).use { npzf ->
            assertArrayEquals(byteArrayOf(1, 2, 3, 4), npzf["x_i1"].asByteArray())
            assertArrayEquals(shortArrayOf(1, 2, 3, 4, 4096), npzf["x_i2"].asShortArray())
            assertArrayEquals(intArrayOf(1, 2, 3, 4), npzf["x_u4"].asIntArray())
            assertArrayEquals(longArrayOf(1, 2, 3, 4, 65536), npzf["x_i8"].asLongArray())

            assertArrayEquals(floatArrayOf(1.0f, 2.0f, 3.0f, 4.0f),
                              npzf["x_f4"].asFloatArray(),
                              Math.ulp(1.0f))
            assertArrayEquals(doubleArrayOf(1.0, 2.0, 3.0, 4.0),
                              npzf["x_f8"].asDoubleArray(),
                              Math.ulp(1.0))

            assertArrayEquals(booleanArrayOf(true, true, true, false),
                              npzf["x_b"].asBooleanArray())

            assertArrayEquals(arrayOf("aha", "hah"), npzf["x_S3"].asStringArray())
        }
    }
}

@RunWith(Parameterized::class)
class NpzFileWriteReadTest(private val order: ByteOrder) {
    @Test fun basic() {
        withTempFile("test", ".npz") { path ->
            val booleans = booleanArrayOf(true, true, true, false)
            val ints = intArrayOf(1, 2, 3, 4)
            val shorts = shortArrayOf(5, 6, 7, 8)
            val floats = floatArrayOf(9.0f, 10.0f, 11.0f, 12.0f)
            val doubles = doubleArrayOf(13.0, 14.0, 15.0, 16.0)
            val longs = longArrayOf(17, 18, 19, 20)
            NpzFile.write(path).use {
                it.write("x_b", booleans)
                it.write("x_i4", ints, order = order)
                it.write("x_i2", shorts, order = order)
                it.write("x_i8", longs, order = order)
                it.write("x_f4", floats, order = order)
                it.write("x_f8", doubles, order = order)
            }

            NpzFile.read(path).use { npzf ->
                assertArrayEquals(booleans, npzf["x_b"].asBooleanArray())
                assertArrayEquals(ints, npzf["x_i4"].asIntArray())
                assertArrayEquals(shorts, npzf["x_i2"].asShortArray())
                assertArrayEquals(longs, npzf["x_i8"].asLongArray())
                assertArrayEquals(floats, npzf["x_f4"].asFloatArray(), 0.0f)
                assertArrayEquals(doubles, npzf["x_f8"].asDoubleArray(), 0.0)
            }
        }
    }

    @Test fun nested() {
        withTempFile("test", ".npz") { path ->
            NpzFile.write(path).use {
                it.write("foo/bar/baz/x_i4", intArrayOf(1, 2, 3, 4), order = order)
            }

            NpzFile.read(path).use { npzf ->
                assertArrayEquals(intArrayOf(1, 2, 3, 4),
                                  npzf["foo/bar/baz/x_i4"].asIntArray())
            }
        }
    }

    @Test fun nd() {
        withTempFile("test", ".npz") { path ->
            NpzFile.write(path).use {
                it.write("x_i4", intArrayOf(1, 2, 3, 4), shape = intArrayOf(2, 2),
                         order = order)
            }

            NpzFile.read(path).use { npzf ->
                assertArrayEquals(intArrayOf(2, 2), npzf["x_i4"].shape)
                assertArrayEquals(intArrayOf(1, 2, 3, 4),
                                  npzf["x_i4"].asIntArray())
            }
        }
    }

    @Test fun unaligned() {
        withTempFile("test", ".npz") { path ->
            val data = IntArray(65536) { it }
            NpzFile.write(path).use {
                it.write("x_i4", data, order = order)
            }

            NpzFile.read(path).use { npzf ->
                assertArrayEquals(data, npzf.get("x_i4", step = 123).asIntArray())
            }
        }
    }

    @Test fun random() {
        val r = Random()
        val maxSize = 65535
        for (i in 0 until 10) {
            val data = Array(r.nextInt(maxSize)) {
                // Random non-empty "ACGT" strings of length no more than 12.
                val alphabet = "ACGT"
                val b = ByteArray(r.nextInt(12) + 1) { alphabet[r.nextInt(4)].toByte() }
                String(b)
            }

            withTempFile("test", ".npz") { path ->
                NpzFile.write(path).use { it.write("data", data) }

                NpzFile.read(path).use {
                    assertArrayEquals(data, it["data"].asStringArray())
                }
            }
        }
    }

    companion object {
        @JvmStatic
        @Parameters(name = "{0}")
        fun `data`(): Collection<Any> = listOf(ByteOrder.BIG_ENDIAN,
                                               ByteOrder.LITTLE_ENDIAN)
    }
}


src/main/kotlin/org/jetbrains/bio/npy/Npy.kt
package org.jetbrains.bio.npy

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel
import java.nio.channels.FileChannel.MapMode
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.*

/**
 * A file in NPY format.
 *
 * Currently unsupported types:
 *
 *   * unsigned integral types (treated as signed)
 *   * bit field,
 *   * complex,
 *   * object,
 *   * Unicode
 *   * void*
 *   * intersections aka types for structured arrays.
 *
 * See http://docs.scipy.org/doc/numpy-dev/neps/npy-format.html
 */
object NpyFile {
    /**
     * NPY file header.
     *
     * Presently NumPy implements two version of the NPY format: 1.0 and 2.0.
     * The difference between the two is the maximum size of the NPY header.
     * Version 1.0 requires it to be <=2**16 while version 2.0 allows <=2**32.
     *
     * The appropriate NPY format is chosen automatically based on the
     * header size.
     */
    internal data class Header(
        val order: ByteOrder? = null,
        val type: Char,
        val bytes: Int,
        val shape: IntArray
    ) {
        /** Major version number. */
        val major: Int
        /** Minor version number. */
        private val minor: Int = 0
        /** Meta-data formatted as a Python dict and 16-byte-padded. */
        private val meta: String

        /** Header size in bytes. */
        private val size: Int

        init {
            val metaUnpadded = StringJoiner(", ", "{", "}")
                .add("'descr': '${order.toChar()}$type$bytes'")
                .add("'fortran_order': False")
                .add("'shape': (${shape.joinToString(",")}, )")
                .toString()

            // According to the spec the total meta size should be
            // evenly divisible by 16 for alignment purposes. +1 here
            // accounts for the newline.
            // XXX despite the fact that the HEADER_LEN is 4 bytes in
            //     NPY2.0 the padding is always computed assuming 2 bytes.
            val totalUnpadded = MAGIC.size + 2 + java.lang.Short.BYTES +
                    metaUnpadded.length + 1
            val padding = 16 - totalUnpadded % 16

            var total = totalUnpadded + padding
            if (total <= NPY_10_20_SIZE_BOUNDARY) {
                major = 1
            } else {
                total += 2  // fix for the XXX above.
                major = 2
            }

            meta = metaUnpadded + " ".repeat(padding) + '\n'
            size = total
        }

        /** Allocates a [ByteBuffer] for this header. */
        internal fun allocate() = ByteBuffer.allocateDirect(size).apply {
            order(ByteOrder.LITTLE_ENDIAN)
            put(MAGIC)
            put(major.toByte())
            put(minor.toByte())

            when (major to minor) {
                1 to 0 -> putShort(meta.length.toShort())
                2 to 0 -> putInt(meta.length)
            }

            put(meta.toByteArray(Charsets.US_ASCII))

            rewind()
        }

        override fun equals(other: Any?) = when {
            this === other -> true
            other == null || other !is Header -> false
            else -> {
                order == other.order &&
                        type == other.type && bytes == other.bytes &&
                        shape.contentEquals(other.shape)
            }
        }

        override fun hashCode() = Objects.hash(order, type, bytes, Arrays.hashCode(shape))

        companion object {
            /** Each NPY file *must* start with this byte sequence. */
            internal val MAGIC = byteArrayOf(0x93.toByte()) + "NUMPY".toByteArray()
            /** Maximum byte size of the header to be written as NPY1.0. */
            // XXX this is a var only for testing purposes.
            internal var NPY_10_20_SIZE_BOUNDARY = 65535

            @Suppress("unchecked_cast", "UsePropertyAccessSyntax")
            // property access syntax is misleading for the buffer "get" operations
            fun read(input: ByteBuffer) = with(input.order(ByteOrder.LITTLE_ENDIAN)) {
                val buf = ByteArray(6)
                get(buf)
                check(MAGIC.contentEquals(buf)) { "bad magic: ${String(buf)}" }

                val major = get().toInt()
                val minor = get().toInt()
                val size = when (major to minor) {
                    1 to 0 -> getShort().toInt()
                    2 to 0 -> getInt()
                    else -> error("unsupported version: $major.$minor")
                }

                val header = ByteArray(size)
                get(header)

                val s = String(header)
                val meta = parseDict(s)
                val type = meta["descr"] as String
                check(!(meta["fortran_order"] as Boolean)) {
                    "Fortran-contiguous arrays are not supported"
                }

                val shape = (meta["shape"] as List<Int>).toIntArray()
                val order = type[0].toByteOrder()
                Header(order = order, type = type[1], bytes = type.substring(2).toInt(), shape = shape)
            }
        }
    }

    /**
     * Reads an array in NPY format from a given path.
     *
     * The caller is responsible for coercing the resulting array to
     * an appropriate type via [NpyArray] methods.
     */
    @JvmStatic
    fun read(path: Path, step: Int = Int.MAX_VALUE): NpyArray {
        return FileChannel.open(path).use {
            var remaining = Files.size(path)
            var chunk = ByteBuffer.allocate(0)
            read(generateSequence {
                // Make sure we don't miss any unaligned bytes.
                remaining += chunk.remaining()
                if (remaining == 0L) {
                    null
                } else {
                    val offset = Files.size(path) - remaining
                    chunk = it.map(
                        MapMode.READ_ONLY, offset,
                        if (remaining > step) step.toLong() else remaining)

                    remaining -= chunk.capacity()
                    chunk
                }
            })
        }
    }

    internal fun read(chunks: Sequence<ByteBuffer>): NpyArray {
        // XXX we have to make it peeking, because otherwise
        //     the first chunk would be gone.
        val iterator = PeekingIterator(chunks.iterator())
        val header = Header.read(iterator.peek())
        val size = header.shape.reduce { a, b -> a * b }
        val merger = when (header.type) {
            'b' -> when (header.bytes) {
                1 -> BooleanArrayMerger(size)
                else -> unexpectedByteNumber(header.type, header.bytes)
            }
            'u', 'i' -> when (header.bytes) {
                1 -> ByteArrayMerger(size)
                2 -> ShortArrayMerger(size)
                4 -> IntArrayMerger(size)
                8 -> LongArrayMerger(size)
                else -> unexpectedByteNumber(header.type, header.bytes)
            }
            'f' -> when (header.bytes) {
                4 -> FloatArrayMerger(size)
                8 -> DoubleArrayMerger(size)
                else -> unexpectedByteNumber(header.type, header.bytes)
            }
            'S' -> StringArrayMerger(size, header.bytes)
            else -> unexpectedHeaderType(header.type)
        }

        for (chunk in iterator) {
            header.order?.let { chunk.order(it) }
            merger(chunk)
        }

        return NpyArray(merger.result(), header.shape)
    }

    /**
     * Writes an array in NPY format to a given path.
     */
    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: BooleanArray,
        shape: IntArray = intArrayOf(data.size)) {
        write(path, allocate(data, shape))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: ByteArray,
        shape: IntArray = intArrayOf(data.size)) {
        write(path, allocate(data, shape))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: ShortArray,
        shape: IntArray = intArrayOf(data.size),
        order: ByteOrder = ByteOrder.nativeOrder()) {
        write(path, allocate(data, shape, order))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: IntArray,
        shape: IntArray = intArrayOf(data.size),
        order: ByteOrder = ByteOrder.nativeOrder()) {
        write(path, allocate(data, shape, order))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: LongArray,
        shape: IntArray = intArrayOf(data.size),
        order: ByteOrder = ByteOrder.nativeOrder()) {
        write(path, allocate(data, shape, order))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: FloatArray,
        shape: IntArray = intArrayOf(data.size),
        order: ByteOrder = ByteOrder.nativeOrder()) {
        write(path, allocate(data, shape, order))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: DoubleArray,
        shape: IntArray = intArrayOf(data.size),
        order: ByteOrder = ByteOrder.nativeOrder()) {
        write(path, allocate(data, shape, order))
    }

    @JvmOverloads
    @JvmStatic
    fun write(path: Path, data: Array<String>,
        shape: IntArray = intArrayOf(data.size)) {
        write(path, allocate(data, shape))
    }

    private fun write(path: Path, chunks: Sequence<ByteBuffer>) {
        FileChannel.open(path,
            StandardOpenOption.WRITE,
            StandardOpenOption.CREATE).use {

            for (chunk in chunks) {
                while (chunk.hasRemaining()) {
                    it.write(chunk)
                }
            }

            it.truncate(it.position())
        }
    }

    internal fun allocate(data: BooleanArray, shape: IntArray): Sequence<ByteBuffer> {
        val header = Header(order = null, type = 'b', bytes = 1, shape = shape)
        return sequenceOf(header.allocate()) + BooleanArrayChunker(data)
    }

    internal fun allocate(data: ByteArray, shape: IntArray): Sequence<ByteBuffer> {
        val header = Header(order = null, type = 'i', bytes = 1, shape = shape)
        return sequenceOf(header.allocate()) + ByteBuffer.wrap(data)
    }

    internal fun allocate(data: ShortArray, shape: IntArray,
        order: ByteOrder): Sequence<ByteBuffer> {
        val header = Header(order = order, type = 'i',
            bytes = java.lang.Short.BYTES, shape = shape)
        return sequenceOf(header.allocate()) + ShortArrayChunker(data, order)
    }

    internal fun allocate(data: IntArray, shape: IntArray,
        order: ByteOrder): Sequence<ByteBuffer> {
        val header = Header(order = order, type = 'i',
            bytes = Integer.BYTES, shape = shape)
        return sequenceOf(header.allocate()) + IntArrayChunker(data, order)
    }

    internal fun allocate(data: LongArray, shape: IntArray,
        order: ByteOrder): Sequence<ByteBuffer> {
        val header = Header(order = order, type = 'i',
            bytes = java.lang.Long.BYTES, shape = shape)
        return sequenceOf(header.allocate()) + LongArrayChunker(data, order)
    }

    internal fun allocate(data: FloatArray, shape: IntArray,
        order: ByteOrder): Sequence<ByteBuffer> {
        val header = Header(order = order, type = 'f',
            bytes = java.lang.Float.BYTES, shape = shape)
        return sequenceOf(header.allocate()) + FloatArrayChunker(data, order)
    }

    internal fun allocate(data: DoubleArray, shape: IntArray,
        order: ByteOrder): Sequence<ByteBuffer> {
        val header = Header(order = order, type = 'f',
            bytes = java.lang.Double.BYTES, shape = shape)
        return sequenceOf(header.allocate()) + DoubleArrayChunker(data, order)
    }

    internal fun allocate(data: Array<String>, shape: IntArray): Sequence<ByteBuffer> {
        val bytes = data.asSequence().map { it.length }.maxOrNull() ?: 0
        val header = Header(order = null, type = 'S', bytes = bytes, shape = shape)
        return sequenceOf(header.allocate()) + StringArrayChunker(data)
    }
}

/** A wrapper for NPY array data. */
class NpyArray(
    /** Array data. */
    val data: Any,
    /** Array dimensions. */
    val shape: IntArray) {

    fun asBooleanArray() = data as BooleanArray

    fun asByteArray() = data as ByteArray

    fun asShortArray() = data as ShortArray

    fun asIntArray() = data as IntArray

    fun asLongArray() = data as LongArray

    fun asFloatArray() = data as FloatArray

    fun asDoubleArray() = data as DoubleArray

    @Suppress("unchecked_cast")
    fun asStringArray() = data as Array<String>

    override fun toString() = StringJoiner(", ", "NpyArray{", "}")
        .add("data=" + arrayOf(data).contentDeepToString()
            .removeSurrounding("[", "]"))
        .add("shape=" + shape.contentToString())
        .toString()
}

private fun Char.toByteOrder() = when (this) {
    '<' -> ByteOrder.LITTLE_ENDIAN
    '>' -> ByteOrder.BIG_ENDIAN
    '|' -> null
    else -> throw IllegalStateException("unexpected byte order char $this")
}

private fun ByteOrder?.toChar() = when (this) {
    ByteOrder.LITTLE_ENDIAN -> '<'
    ByteOrder.BIG_ENDIAN -> '>'
    null -> '|'
    else -> error(this)
}